pkg/storegateway/indexheader/thread.go

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cpg1111/threadpool-go/blob/master/thread.go
+// Provenance-includes-license: MIT
+// Provenance-includes-copyright: Christian Grabowski
+
+package indexheader
+
+import (
+	"runtime"
+)
+
+type execResult struct {
+	value interface{}
+	err   error
+}
+
+type osThread struct {
+	// stopping is a request for this thread to stop running.
+	stopping chan struct{}
+
+	// stopped is confirmation that the thread has stopped running.
+	stopped chan struct{}
+
+	// call is used to submit closures for this thread to execute.
+	call chan func() (interface{}, error)
+
+	// res is used to send the results of running closures back to callers.
+	res chan execResult
+}
+
+func newOSThread(stopping chan struct{}) *osThread {
+	return &osThread{
+		stopping: stopping,
+		stopped:  make(chan struct{}),
+
+		call: make(chan func() (interface{}, error)),
+		res:  make(chan execResult),
+	}
+}
+
+func (o *osThread) run() {
+	runtime.LockOSThread()
+
+	defer func() {
+		runtime.UnlockOSThread()
+		close(o.stopped)
+	}()
+
+	for {
+		select {
+		case <-o.stopping:
+			return
+		case fn := <-o.call:
+			val, err := fn()
+			o.res <- execResult{value: val, err: err}
+		}
+	}
+}
+
+func (o *osThread) start() {
+	go o.run()
+}
+
+// join waits for the thread to complete any tasks it was executing after the
+// stopping channel is closed by the parent thread pool. This allows callers to
+// ensure the thread has completed executing which is useful in tests to verify
+// no goroutines leak.
+func (o *osThread) join() {
+	<-o.stopped
+}
+
+func (o *osThread) execute(fn func() (interface{}, error)) (interface{}, error) {
+	select {
+	case <-o.stopping:
+		// Make sure the pool (and hence this thread) hasn't been stopped before
+		// submitting our function to run.
+		return nil, ErrPoolStopped
+	case o.call <- fn:
+		res := <-o.res
+		return res.value, res.err
+	}
```

Comment:
Just as a safety measure, I'd consider adding a default clause here:
```go
default:
        panic("either called on a non-running thread or used same thread for two concurrent calls")
```

This should never happen, but if we end up in this situation, it's better to panic and restart than to have an unresponsive application saying nothing.

Reply:
OK, will do.

Reply:
Actually, it doesn't seem like we can do this. I suspect this is because the write to `o.call` would block because the `o.start` method has not yet started consuming entries from `o.call` when run in this test and thus the `default` case gets executed. Without the `default` case, the  write to `o.call` blocks until the thread reads the value in `o.start`.

```
--- FAIL: TestOSThread_Call (0.01s)
    --- FAIL: TestOSThread_Call/run_by_thread (0.01s)
panic: either called on a non-running thread or used same thread for two concurrent calls [recovered]
        panic: either called on a non-running thread or used same thread for two concurrent calls

goroutine 8 [running]:
testing.tRunner.func1.2({0x982820, 0xc24f38})
        /usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1212 +0x218
panic({0x982820, 0xc24f38})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/grafana/mimir/pkg/storegateway/indexheader.(*osThread).execute(0xc00007d400, 0xc00000ec60)
        /go/src/github.com/grafana/mimir/pkg/storegateway/indexheader/thread.go:87 +0x108
github.com/grafana/mimir/pkg/storegateway/indexheader.TestOSThread_Call.func1(0xc0000b9040)
        /go/src/github.com/grafana/mimir/pkg/storegateway/indexheader/thread_test.go:25 +0x185
testing.tRunner(0xc0000b9040, 0xa77c48)
        /usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1306 +0x35a
```

Reply:
Oh, I see. Certainly it's hard-to-impossible to fix this (even if that goroutine is started, it may not be listening on the for-loop yet). We could set a timer here, but that would be too much hassle for something that shouldn't happen.